finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,645 @@
+            .await;
+        assert_eq!(message.included_data(), test_data);
+    }
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn too_long_branch_message_does_not_go_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 2].hash());
+
+    let blocks_branch = blocks[0..((MAX_DATA_BRANCH_LEN + 1) as usize)].to_vec();
+    let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+    test_handler.send_data(test_data.clone());
+    test_handler
+        .assert_no_message_out("Data Store let through a too long message")
+        .await;
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn branch_with_not_finalized_ancestor_correctly_handled() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    let blocks_branch = blocks[1..2].to_vec();
+    let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+    test_handler.send_data(test_data.clone());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not finalized ancestor")
+        .await;
+
+    // After the block gets finalized the message should be let through
+    test_handler.finalize_block(&blocks[0].hash());
+    let message = test_handler
+        .assert_message_out("Did not receive message from Data Store")
+        .await;
+    assert_eq!(message.included_data(), test_data);
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn correct_messages_go_through_with_late_import() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.import_branch(blocks, false).await;
+
+    for _ in 1..=MAX_DATA_BRANCH_LEN {
+        let _message = test_handler
```

Comment:
I think it's better to remove `let _message = ` here and below. Just having it here suggests we want to keep it alive for some reason?